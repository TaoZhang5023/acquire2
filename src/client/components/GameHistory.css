.root {
    background-color: #c0c0ff;
    height: 1px;
    overflow-y: scroll;
}

.turnBegan {
    border-bottom: none;
    border-left: none;
    border-right: none;
    border-top: 1px solid #000000;
    margin: 0;
    padding: 0;
    text-align: center;
    & legend {
        padding: 0 5px;
    }
}

.move {
    padding: 0 2px;
}

.move:hover {
    cursor: pointer;
}

.move:not(.selected):hover {
    background-color: #ffefc0;
}

.move.selected {
    background-color: #ffd0c0;
}
