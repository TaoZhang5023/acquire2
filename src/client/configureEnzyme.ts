import { configure } from 'enzyme';
import * as EnzymeAdapter from 'enzyme-adapter-react-16';

export function configureEnzyme() {
    configure({ adapter: new EnzymeAdapter() });
}
