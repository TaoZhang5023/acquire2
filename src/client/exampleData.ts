import { List } from 'immutable';
import { defaultMoveDataHistory } from '../common/defaults';
import { GameBoardType, GameHistoryMessage, GameMode, PlayerArrangementMode } from '../common/enums';
import { Game } from '../common/game';
import { ActionDisposeOfShares } from '../common/gameActions/disposeOfShares';
import { ActionGameOver } from '../common/gameActions/gameOver';
import { ActionPlayTile } from '../common/gameActions/playTile';
import { ActionPurchaseShares } from '../common/gameActions/purchaseShares';
import { ActionSelectChainToDisposeOfNext } from '../common/gameActions/selectChainToDisposeOfNext';
import { ActionSelectMergerSurvivor } from '../common/gameActions/selectMergerSurvivor';
import { ActionSelectNewChain } from '../common/gameActions/selectNewChain';
import { ActionStartGame } from '../common/gameActions/startGame';
import { getNewTileBag } from '../common/helpers';
import { allChains } from './helpers';

export function getDummyGameForGetGameHistory() {
    const game = new Game(GameMode.Singles4, PlayerArrangementMode.ExactOrder, getNewTileBag(), List([2, 3, 5, 8]), List(['Tim', 'Rita', 'Dad', 'Mom']), 8, 3);
    game.doGameAction([], null);
    game.moveDataHistory = defaultMoveDataHistory;

    let moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.TurnBegan, 0, []);
    moveData.timestamp = 1524896229792;
    game.endCurrentMove();

    moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.DrewPositionTile, 1, [21]);
    game.endCurrentMove();

    moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.StartedGame, 2, []);
    game.endCurrentMove();

    moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.DrewTile, 3, [100]);
    game.endCurrentMove();

    moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.HasNoPlayableTile, 0, []);
    game.endCurrentMove();

    moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.PlayedTile, 1, [40]);
    game.endCurrentMove();

    moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.FormedChain, 2, [0]);
    game.endCurrentMove();

    moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.MergedChains, 3, [[1, 2]]);
    moveData.addGameHistoryMessage(GameHistoryMessage.MergedChains, 0, [[3, 4, 5]]);
    moveData.addGameHistoryMessage(GameHistoryMessage.MergedChains, 1, [[0, 1, 2, 6]]);
    game.endCurrentMove();

    moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.SelectedMergerSurvivor, 2, [3]);
    game.endCurrentMove();

    moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.SelectedChainToDisposeOfNext, 3, [4]);
    game.endCurrentMove();

    moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.ReceivedBonus, 0, [5, 25]);
    game.endCurrentMove();

    moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.DisposedOfShares, 1, [6, 2, 3]);
    game.endCurrentMove();

    moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.CouldNotAffordAnyShares, 2, []);
    game.endCurrentMove();

    moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.PurchasedShares, 3, [[]]);
    moveData.addGameHistoryMessage(GameHistoryMessage.PurchasedShares, 0, [[[0, 3]]]);
    moveData.addGameHistoryMessage(GameHistoryMessage.PurchasedShares, 1, [[[1, 2], [2, 1]]]);
    moveData.addGameHistoryMessage(GameHistoryMessage.PurchasedShares, 2, [[[3, 1], [4, 1], [5, 1]]]);
    game.endCurrentMove();

    moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.DrewLastTile, 3, []);
    game.endCurrentMove();

    moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.ReplacedDeadTile, 0, [30]);
    game.endCurrentMove();

    moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.EndedGame, 1, []);
    game.endCurrentMove();

    moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.NoTilesPlayedForEntireRound, null, []);
    game.endCurrentMove();

    moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.AllTilesPlayed, null, []);
    game.endCurrentMove();

    return game;
}

export function getExampleNextGameActionsArray() {
    const game = new Game(
        GameMode.Singles5,
        PlayerArrangementMode.RandomOrder,
        [],
        List([1, 2, 3, 4, 5]),
        List(['Tim', 'Rita', 'Dad', 'Mom', 'REALLY, REALLY, REALLY, REALLY, REALLY LONG NAME']),
        1,
        6,
    );
    return [
        new ActionStartGame(game, 4),
        new ActionStartGame(game, 0),
        new ActionPlayTile(game, 1),
        new ActionSelectNewChain(game, 2, allChains, 107),
        new ActionSelectMergerSurvivor(game, 3, [GameBoardType.Luxor, GameBoardType.Festival, GameBoardType.Continental], 107),
        new ActionSelectChainToDisposeOfNext(game, 0, [GameBoardType.Tower, GameBoardType.American], GameBoardType.Continental),
        new ActionDisposeOfShares(game, 1, GameBoardType.Imperial, GameBoardType.Luxor),
        new ActionPurchaseShares(game, 2),
        new ActionGameOver(game, 3),
    ];
}
