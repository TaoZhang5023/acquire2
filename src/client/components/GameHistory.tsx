import * as style from './GameHistory.scss';

import { List } from 'immutable';
import * as React from 'react';
import { GameHistoryMessage } from '../../common/enums';
import { GameHistoryMessageData, MoveData } from '../../common/game';
import { getHotelNameSpan, getTileString, getUsernameSpan } from '../helpers';
import { GameBoardType } from '../../common/pb';

export interface GameHistoryProps {
  usernames: List<string>;
  moveDataHistory: List<MoveData>;
  selectedMove?: number;
  onMoveClicked: (index: number) => void;
}

export class GameHistory extends React.PureComponent<GameHistoryProps> {
  parentRef: React.RefObject<HTMLDivElement>;
  selectedMoveRef: React.RefObject<HTMLDivElement>;

  constructor(props: GameHistoryProps) {
    super(props);

    this.parentRef = React.createRef();
    this.selectedMoveRef = React.createRef();
  }

  componentDidMount() {
    setTimeout(this.autoScroll, 0);
  }

  componentDidUpdate() {
    setTimeout(this.autoScroll, 0);
  }

  autoScroll = () => {
    const parent = this.parentRef.current;
    const selectedMove = this.selectedMoveRef.current;

    if (parent !== null && selectedMove !== null) {
      const parentScrollTop = parent.scrollTop;
      const parentScrollBottom = parentScrollTop + parent.clientHeight;

      const selectedMoveRelativeOffsetTop = selectedMove.offsetTop - parent.offsetTop;
      const selectedMoveRelativeOffsetBottom = selectedMoveRelativeOffsetTop + selectedMove.clientHeight;

      if (selectedMoveRelativeOffsetTop < parentScrollTop || selectedMove.clientHeight > parent.clientHeight) {
        parent.scrollTop = selectedMoveRelativeOffsetTop;
      } else if (selectedMoveRelativeOffsetBottom > parentScrollBottom) {
        parent.scrollTop = selectedMoveRelativeOffsetBottom - parent.clientHeight;
      }
    }
  };

  render() {
    const { usernames, moveDataHistory, selectedMove, onMoveClicked } = this.props;

    return (
      <div className={style.root} ref={this.parentRef}>
        {moveDataHistory.map((moveData, i) => {
          const optionalProps: { [key: string]: any } = {};
          if (i === selectedMove) {
            optionalProps.ref = this.selectedMoveRef;
          }

          return (
            <div key={i} {...optionalProps}>
              <MoveHistory usernames={usernames} moveData={moveData} moveIndex={i} isSelected={i === selectedMove} onMoveClicked={onMoveClicked} />
            </div>
          );
        })}
      </div>
    );
  }
}

interface MoveHistoryProps {
  usernames: List<string>;
  moveData: MoveData;
  moveIndex: number;
  isSelected: boolean;
  onMoveClicked: (index: number) => void;
}

class MoveHistory extends React.PureComponent<MoveHistoryProps> {
  render() {
    const { usernames, moveData, moveIndex, isSelected, onMoveClicked } = this.props;

    const optionalProps: { [key: string]: any } = {};
    if (moveData.timestamp !== null) {
      const date = new Date(moveData.timestamp);
      optionalProps.title = date.toString();
    }

    return (
      <div className={style.move + (isSelected ? ` ${style.selected}` : '')} {...optionalProps} onClick={() => onMoveClicked(moveIndex)}>
        {moveData.gameHistoryMessages.map((ghmd, index) => {
          const username = ghmd.playerID === null ? '' : usernames.get(ghmd.playerID)!;
          return gameHistoryMessageHandlerLookup.get(ghmd.gameHistoryMessage)!(index, username, ghmd);
        })}
      </div>
    );
  }
}

const gameHistoryMessageHandlerLookup = new Map<GameHistoryMessage, (key: number, username: string, ghmd: GameHistoryMessageData) => JSX.Element>([
  [
    GameHistoryMessage.TurnBegan,
    (key, username) => (
      <div key={key}>
        <fieldset className={style.turnBegan}>
          <legend>{getUsernameSpan(username)}</legend>
        </fieldset>
      </div>
    ),
  ],
  [
    GameHistoryMessage.DrewPositionTile,
    (key, username, ghmd) => (
      <div key={key}>
        {getUsernameSpan(username)} drew position tile {getTileString(ghmd.parameters[0])}.
      </div>
    ),
  ],
  [GameHistoryMessage.StartedGame, (key, username) => <div key={key}>{getUsernameSpan(username)} started the game.</div>],
  [
    GameHistoryMessage.DrewTile,
    (key, username, ghmd) => (
      <div key={key}>
        {getUsernameSpan(username)} drew tile {getTileString(ghmd.parameters[0])}.
      </div>
    ),
  ],
  [GameHistoryMessage.HasNoPlayableTile, (key, username) => <div key={key}>{getUsernameSpan(username)} has no playable tile.</div>],
  [
    GameHistoryMessage.PlayedTile,
    (key, username, ghmd) => (
      <div key={key}>
        {getUsernameSpan(username)} played tile {getTileString(ghmd.parameters[0])}.
      </div>
    ),
  ],
  [
    GameHistoryMessage.FormedChain,
    (key, username, ghmd) => (
      <div key={key}>
        {getUsernameSpan(username)} formed {getHotelNameSpan(ghmd.parameters[0])}.
      </div>
    ),
  ],
  [
    GameHistoryMessage.MergedChains,
    (key, username, ghmd) => {
      const chains: number[] = ghmd.parameters[0];
      if (chains.length === 2) {
        return (
          <div key={key}>
            {getUsernameSpan(username)} merged {getHotelNameSpan(chains[0])} and {getHotelNameSpan(chains[1])}.
          </div>
        );
      } else if (chains.length === 3) {
        return (
          <div key={key}>
            {getUsernameSpan(username)} merged {getHotelNameSpan(chains[0])}, {getHotelNameSpan(chains[1])}, and {getHotelNameSpan(chains[2])}.
          </div>
        );
      } else {
        return (
          <div key={key}>
            {getUsernameSpan(username)} merged {getHotelNameSpan(chains[0])}, {getHotelNameSpan(chains[1])}, {getHotelNameSpan(chains[2])}, and{' '}
            {getHotelNameSpan(chains[3])}.
          </div>
        );
      }
    },
  ],
  [
    GameHistoryMessage.SelectedMergerSurvivor,
    (key, username, ghmd) => (
      <div key={key}>
        {getUsernameSpan(username)} selected {getHotelNameSpan(ghmd.parameters[0])} as merger survivor.
      </div>
    ),
  ],
  [
    GameHistoryMessage.SelectedChainToDisposeOfNext,
    (key, username, ghmd) => (
      <div key={key}>
        {getUsernameSpan(username)} selected {getHotelNameSpan(ghmd.parameters[0])} as chain to dispose of next.
      </div>
    ),
  ],
  [
    GameHistoryMessage.ReceivedBonus,
    (key, username, ghmd) => (
      <div key={key}>
        {getUsernameSpan(username)} received a ${ghmd.parameters[1] * 100} {getHotelNameSpan(ghmd.parameters[0])} bonus.
      </div>
    ),
  ],
  [
    GameHistoryMessage.DisposedOfShares,
    (key, username, ghmd) => (
      <div key={key}>
        {getUsernameSpan(username)} traded {ghmd.parameters[1]} and sold {ghmd.parameters[2]} {getHotelNameSpan(ghmd.parameters[0])} shares.
      </div>
    ),
  ],
  [GameHistoryMessage.CouldNotAffordAnyShares, (key, username) => <div key={key}>{getUsernameSpan(username)} could not afford any shares.</div>],
  [
    GameHistoryMessage.PurchasedShares,
    (key, username, ghmd) => {
      const chainsAndCounts: [GameBoardType, number][] = ghmd.parameters[0];
      if (chainsAndCounts.length === 0) {
        return <div key={key}>{getUsernameSpan(username)} purchased nothing.</div>;
      } else if (chainsAndCounts.length === 1) {
        return (
          <div key={key}>
            {getUsernameSpan(username)} purchased {chainsAndCounts[0][1]} {getHotelNameSpan(chainsAndCounts[0][0])}.
          </div>
        );
      } else if (chainsAndCounts.length === 2) {
        return (
          <div key={key}>
            {getUsernameSpan(username)} purchased {chainsAndCounts[0][1]} {getHotelNameSpan(chainsAndCounts[0][0])} and {chainsAndCounts[1][1]}{' '}
            {getHotelNameSpan(chainsAndCounts[1][0])}.
          </div>
        );
      } else {
        return (
          <div key={key}>
            {getUsernameSpan(username)} purchased {chainsAndCounts[0][1]} {getHotelNameSpan(chainsAndCounts[0][0])}, {chainsAndCounts[1][1]}{' '}
            {getHotelNameSpan(chainsAndCounts[1][0])}, and {chainsAndCounts[2][1]} {getHotelNameSpan(chainsAndCounts[2][0])}.
          </div>
        );
      }
    },
  ],
  [GameHistoryMessage.DrewLastTile, (key, username) => <div key={key}>{getUsernameSpan(username)} drew the last tile from the tile bag.</div>],
  [
    GameHistoryMessage.ReplacedDeadTile,
    (key, username, ghmd) => (
      <div key={key}>
        {getUsernameSpan(username)} replaced dead tile {getTileString(ghmd.parameters[0])}.
      </div>
    ),
  ],
  [GameHistoryMessage.EndedGame, (key, username) => <div key={key}>{getUsernameSpan(username)} ended the game.</div>],
  [GameHistoryMessage.NoTilesPlayedForEntireRound, (key) => <div key={key}>No tiles played for an entire round. Game end forced.</div>],
  [GameHistoryMessage.AllTilesPlayed, (key) => <div key={key}>All tiles have been played. Game end forced.</div>],
]);
