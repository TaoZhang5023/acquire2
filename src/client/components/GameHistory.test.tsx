import { mount } from 'enzyme';
import { List } from 'immutable';
import * as React from 'react';
import * as renderer from 'react-test-renderer';

import { configureEnzyme } from '../configureEnzyme';
import { getDummyGameForGetGameHistory } from '../exampleData';
import { GameHistory } from './GameHistory';

configureEnzyme();

describe('GameHistory', () => {
    const onMoveClicked = jest.fn();
    const game = getDummyGameForGetGameHistory();
    const jsx = (
        <GameHistory
            usernames={List(['Tim', 'Rita', 'Dad', 'Mom'])}
            moveDataHistory={game.moveDataHistory}
            selectedMove={1}
            width={600}
            height={300}
            onMoveClicked={onMoveClicked}
        />
    );

    it('renders correctly', () => {
        const component = renderer.create(jsx);
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('works correctly', () => {
        const gameHistory = mount(jsx);

        gameHistory
            .find('.move')
            .at(3)
            .simulate('click');

        expect(onMoveClicked.mock.calls.length).toBe(1);
        expect(onMoveClicked.mock.calls[0]).toEqual([3]);
    });
});
