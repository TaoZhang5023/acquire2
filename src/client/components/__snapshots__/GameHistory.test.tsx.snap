// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameHistory renders correctly 1`] = `
<div
  className="root"
>
  <div>
    <div
      className="move"
      onClick={[Function]}
      title="new Date(1524896229792).toString()"
    >
      <div>
        <fieldset
          className="turnBegan"
        >
          <legend>
            <span
              className="username"
            >
              Tim
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
  <div>
    <div
      className="move selected"
      onClick={[Function]}
    >
      <div>
        <span
          className="username"
        >
          Rita
        </span>
         drew position tile 
        3D
        .
      </div>
    </div>
  </div>
  <div>
    <div
      className="move"
      onClick={[Function]}
    >
      <div>
        <span
          className="username"
        >
          Dad
        </span>
         started the game.
      </div>
    </div>
  </div>
  <div>
    <div
      className="move"
      onClick={[Function]}
    >
      <div>
        <span
          className="username"
        >
          Mom
        </span>
         drew tile 
        12B
        .
      </div>
    </div>
  </div>
  <div>
    <div
      className="move"
      onClick={[Function]}
    >
      <div>
        <span
          className="username"
        >
          Tim
        </span>
         has no playable tile.
      </div>
    </div>
  </div>
  <div>
    <div
      className="move"
      onClick={[Function]}
    >
      <div>
        <span
          className="username"
        >
          Rita
        </span>
         played tile 
        5E
        .
      </div>
    </div>
  </div>
  <div>
    <div
      className="move"
      onClick={[Function]}
    >
      <div>
        <span
          className="username"
        >
          Dad
        </span>
         formed 
        <span
          className="luxor"
        >
          Luxor
        </span>
        .
      </div>
    </div>
  </div>
  <div>
    <div
      className="move"
      onClick={[Function]}
    >
      <div>
        <span
          className="username"
        >
          Mom
        </span>
         merged 
        <span
          className="tower"
        >
          Tower
        </span>
         and 
        <span
          className="american"
        >
          American
        </span>
        .
      </div>
      <div>
        <span
          className="username"
        >
          Tim
        </span>
         merged 
        <span
          className="festival"
        >
          Festival
        </span>
        , 
        <span
          className="worldwide"
        >
          Worldwide
        </span>
        , and 
        <span
          className="continental"
        >
          Continental
        </span>
        .
      </div>
      <div>
        <span
          className="username"
        >
          Rita
        </span>
         merged 
        <span
          className="luxor"
        >
          Luxor
        </span>
        , 
        <span
          className="tower"
        >
          Tower
        </span>
        , 
        <span
          className="american"
        >
          American
        </span>
        , and
         
        <span
          className="imperial"
        >
          Imperial
        </span>
        .
      </div>
    </div>
  </div>
  <div>
    <div
      className="move"
      onClick={[Function]}
    >
      <div>
        <span
          className="username"
        >
          Dad
        </span>
         selected 
        <span
          className="festival"
        >
          Festival
        </span>
         as merger survivor.
      </div>
    </div>
  </div>
  <div>
    <div
      className="move"
      onClick={[Function]}
    >
      <div>
        <span
          className="username"
        >
          Mom
        </span>
         selected 
        <span
          className="worldwide"
        >
          Worldwide
        </span>
         as chain to dispose of next.
      </div>
    </div>
  </div>
  <div>
    <div
      className="move"
      onClick={[Function]}
    >
      <div>
        <span
          className="username"
        >
          Tim
        </span>
         received a $
        2500
         
        <span
          className="continental"
        >
          Continental
        </span>
         bonus.
      </div>
    </div>
  </div>
  <div>
    <div
      className="move"
      onClick={[Function]}
    >
      <div>
        <span
          className="username"
        >
          Rita
        </span>
         traded 
        2
         and sold 
        3
         
        <span
          className="imperial"
        >
          Imperial
        </span>
         shares.
      </div>
    </div>
  </div>
  <div>
    <div
      className="move"
      onClick={[Function]}
    >
      <div>
        <span
          className="username"
        >
          Dad
        </span>
         could not afford any shares.
      </div>
    </div>
  </div>
  <div>
    <div
      className="move"
      onClick={[Function]}
    >
      <div>
        <span
          className="username"
        >
          Mom
        </span>
         purchased nothing.
      </div>
      <div>
        <span
          className="username"
        >
          Tim
        </span>
         purchased 
        3
         
        <span
          className="luxor"
        >
          Luxor
        </span>
        .
      </div>
      <div>
        <span
          className="username"
        >
          Rita
        </span>
         purchased 
        2
         
        <span
          className="tower"
        >
          Tower
        </span>
         and 
        1
         
        <span
          className="american"
        >
          American
        </span>
        .
      </div>
      <div>
        <span
          className="username"
        >
          Dad
        </span>
         purchased 
        1
         
        <span
          className="festival"
        >
          Festival
        </span>
        , 
        1
         
        <span
          className="worldwide"
        >
          Worldwide
        </span>
        , and 
        1
         
        <span
          className="continental"
        >
          Continental
        </span>
        .
      </div>
    </div>
  </div>
  <div>
    <div
      className="move"
      onClick={[Function]}
    >
      <div>
        <span
          className="username"
        >
          Mom
        </span>
         drew the last tile from the tile bag.
      </div>
    </div>
  </div>
  <div>
    <div
      className="move"
      onClick={[Function]}
    >
      <div>
        <span
          className="username"
        >
          Tim
        </span>
         replaced dead tile 
        4D
        .
      </div>
    </div>
  </div>
  <div>
    <div
      className="move"
      onClick={[Function]}
    >
      <div>
        <span
          className="username"
        >
          Rita
        </span>
         ended the game.
      </div>
    </div>
  </div>
  <div>
    <div
      className="move"
      onClick={[Function]}
    >
      <div>
        No tiles played for an entire round. Game end forced.
      </div>
    </div>
  </div>
  <div>
    <div
      className="move"
      onClick={[Function]}
    >
      <div>
        All tiles have been played. Game end forced.
      </div>
    </div>
  </div>
</div>
`;
